libraries/AC_SafeRTL/AC_SafeRTL.cpp

```diff
@@ -0,0 +1,573 @@
+        if (HYPOT(current_pos, _path[_last_index]) > _accuracy) {
+            // add the breadcrumb
+            _path[++_last_index] = current_pos;
+
+            log_action(SRTL_POINT_ADD, current_pos);
+
+            // if cleanup algorithms are finished (and therefore not running), reset them
+            if (_simplification_complete) {
+                reset_simplification(false);
+            }
+            if (_pruning_complete) {
+                reset_pruning(false);
+            }
+        }
+    }
+}
+
+/**
+*  Run this method only when preparing to initiate the RTL procedure. Returns a
+*  pointer to the cleaned-up path. Returns nullptr if the cleanup algorithms aren't ready yet.
+*  If this happens, just run this method again a bit later.
+*
+*  Probably best not to run this unless cleanup_ready() is returning true
+*/
+bool AC_SafeRTL::thorough_cleanup()
+{
+    // this should never happen but just in case
+    if (!_active) {
+        return false;
+    }
+
+    // check if we are ready to perform cleanup.  this should be called just before thorough_cleanup just before initiating the RTL
+    if (!_pruning_complete || !_simplification_complete) {
+        return false;
+    }
+
+    // apply simplification
+    zero_points_by_simplification_bitmask();
+
+    // apply pruning, prune every single loop
+    zero_points_by_loops(SAFERTL_PATH_LEN_MAX);
+
+    remove_empty_points();
+
+    // end by resetting the state of the cleanup methods.
+    reset_simplification(true);
+    reset_pruning(true);
+
+    return true;
+}
+
+/**
+*   Returns true if the list is empty after popping this point.
+*/
+bool AC_SafeRTL::pop_point(Vector3f& point)
+{
+    // this should never happen but just in case
+    if (!_active) {
+        return false;
+    }
+
+    if (_last_index == 0) {
+        point = _path[0];
+        return true;
+    }
+    point =  _path[_last_index--];
+    return false;
+}
+
+/**
+*    Simplifies a 3D path, according to the Ramer-Douglas-Peucker algorithm.
+*    Returns the number of items which were removed. end_index is the index of the last element in the path.
+*/
+void AC_SafeRTL::detect_simplifications()
+{
+    if (!_active || _simplification_complete || _last_index < 2) {
+        return;
+    } else if (_simplification_stack_last_index == -1) {  // if not complete but also nothing to do, we must be restarting.
+        // reset to beginning state
+        _simplification_stack[++_simplification_stack_last_index] = start_finish {0, _last_index};
+    }
+    uint32_t start_time = AP_HAL::micros();
+    int16_t start_index, end_index;
+    while (_simplification_stack_last_index >= 0) { // while there is something to do
+        if (AP_HAL::micros() - start_time > SAFERTL_SIMPLIFICATION_TIME_US) {
+            return;
+        }
+
+        start_finish tmp {}; // initialize to zero to suppress warnings
+        tmp = _simplification_stack[_simplification_stack_last_index--];
+        start_index = tmp.start;
```

Comment:
instead i've changed the check so it uses end-index.  it's true that it wastes one assign but it's possible the assignment doesn't even happen after it's all compiled.